### PR TITLE
feat(condo): DOMA-4121 added MessageBatch schema & migration and other

### PR DESCRIPTION
**Pull request by akarjakin**
- Added `notification.MessageBatch` schema
- Added migration
- Added `sendMessageBatch` task
- Added tests for schema and task helpers
- Fixed send-custom-message-batch script to use MessageBatch schema
- Some minor refactoring done

### CODE REVIEW

File: apps/condo/domains/notification/access/MessageBatch.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// @ts-nocheck
```
Comment (dkoviazin) [inline apps/condo/domains/notification/access/MessageBatch.js:1]: // @ts-nocheck

File: apps/condo/domains/notification/tasks/sendMessageBatch.helpers.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+const {
+    SMS_TRANSPORT, EMAIL_TRANSPORT, PUSH_TRANSPORT,
+    CUSTOM_CONTENT_MESSAGE_TYPE,
+} = require('../constants/constants')
+
+const EMAIL_FROM = 'noreply@doma.ai'
+const DATE_FORMAT = 'YYYY-MM-DD'
+const IS_EMAIL_REGEXP = /^\S+@\S+\.\S+$/
+const IS_PHONE_REGEXP = /^\+79\d{9}$/
+const IS_UUID_REGEXP = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i
```
Comment (dkoviazin) [inline apps/condo/domains/notification/tasks/sendMessageBatch.helpers.js:18]: maybe, move all regexps to constants?

IS_PHONE_REGEXP = +7 // Only Russia ?

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotifications.js [hunk 3]
```diff
@@ -118,8 +118,8 @@ const sendBillingReceiptsAddedNotificationsForPeriod = async (receiptsWhere, onL
         const accountsData = receipts.map(receipt => ({
             accountNumber: get(receipt, 'account.number'),
             resident: {
-                unitType: get(receipt, 'account.unitType'),
-                unitName: get(receipt, 'account.unitName'),
+                unitType_i: get(receipt, 'account.unitType'),
```
Comment (dkoviazin) [inline apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotifications.js:121]: accountNumber + house address

File: packages/keystone/tasks.js [hunk 4]
```diff
@@ -173,6 +173,7 @@ function registerTasks (modulesList) {
             Object.values(module).forEach(
                 (task) => {
                     if (task._type !== TASK_TYPE) {
+                        console.error(task)
```
Comment (dkoviazin) [inline packages/keystone/tasks.js:176]: debug

File: apps/condo/domains/notification/tasks/sendMessageBatch.helpers.js (hunk 2 above)
Comment (akarjakin) [reply apps/condo/domains/notification/tasks/sendMessageBatch.helpers.js:18]: by now we support only Russian mobile phones for sms

File: packages/keystone/tasks.js (hunk 4 above)
Comment (akarjakin) [reply packages/keystone/tasks.js:176]: it's useful when you're adding tasks, it'll work only on local env

File: apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotifications.js (hunk 3 above)
Comment (akarjakin) [reply apps/condo/domains/resident/tasks/sendBillingReceiptsAddedNotifications.js:121]: will be fixed in separate PR

**Review by dkoviazin — APPROVED** [review]

**Review by AntonAL — COMMENTED** [review]
Is it really convenient to enter data by hands from console when using `send-custom-message-batch.js` script ?

Maybe to load from YAML file of strict format.

For example:

```
message:
  title: Hello
  body: |
  	This is a custom message
  	With multiple lines
  deepLink: https://example.com/#hello
  phones: |
  	+79991111111
  	+79992222222
  	+79993333333
  emails: |
    anton@example.com
    andrew@example.com
    pavel@example.com
```

This way it's more easier to work, because this file can be prepared by other person, you can store it somewhere, send to somebody etc.

**Review by AntonAL — COMMENTED** [review]
A function `connectKeystone` is duplicated in 4 places.
Maybe it's time to extract and reuse it?

File: apps/condo/domains/common/utils/sleep.js [hunk 5]
```diff
@@ -3,7 +3,7 @@
  */
 const sleep = async (time) => (
     new Promise(resolve => {
-        setTimeout(resolve, time)
+        setTimeout(resolve, time + 1)
```
Comment (AntonAL) [inline apps/condo/domains/common/utils/sleep.js:6]: Maybe a little bit more to be sure?)

File: apps/condo/domains/notification/schema/MessageBatch.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    MESSAGE_BATCH_TYPE_OPTIONS,
+} = require('@condo/domains/notification/constants/constants')
+
+const { sendMessageBatch } = require('../tasks/sendMessageBatch')
+
+const operationForbiddenValidator = ({ fieldPath, addFieldValidationError, operation }) => {
+    if (operation === 'update') addFieldValidationError(`${OPERATION_FORBIDDEN}] Updating ${fieldPath} is forbidden.`)
+}
+
+const MessageBatch = new GQLListSchema('MessageBatch', {
```
Comment (AntonAL) [inline apps/condo/domains/notification/schema/MessageBatch.js:32]: A name `MessageBatch` does not reflects it's primary side effect.

```suggestion
const MessageBatchSendingTask = new GQLListSchema('MessageBatch', {
```

File: apps/condo/domains/notification/schema/MessageBatch.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            options: MESSAGE_BATCH_TYPE_OPTIONS,
+            defaultValue: CUSTOM_CONTENT_MESSAGE_TYPE,
+            isRequired: false,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        title: {
+            schemaDoc: 'Title for a the message to send',
```
Comment (AntonAL) [inline apps/condo/domains/notification/schema/MessageBatch.js:47]: Maybe also provide some details about format. For example, can it be spanned to multiple lines or not, does it supports interpolation etc.
```suggestion
            schemaDoc: 'Common title for messages to be sent',
```

File: apps/condo/domains/notification/schema/MessageBatch.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+            type: Text,
+            isRequired: true,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        message: {
+            schemaDoc: 'Message to send',
```
Comment (AntonAL) [inline apps/condo/domains/notification/schema/MessageBatch.js:56]: The same thing about details.

```suggestion
            schemaDoc: 'Common body for messages to be sent',
```

File: apps/condo/domains/notification/schema/MessageBatch.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            type: Text,
+            isRequired: true,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        deepLink: {
+            schemaDoc: 'Used for push notifications only. Useful for moving a user to proper position within mobile apps',
```
Comment (AntonAL) [inline apps/condo/domains/notification/schema/MessageBatch.js:65]:
```suggestion
            schemaDoc: 'A link to bring user to specified position within a mobile app. Used ONLY for push notifications',
```

File: apps/condo/domains/notification/schema/MessageBatch.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            type: Text,
+            hooks: {
+                validateInput: operationForbiddenValidator,
+            },
+        },
+
+        targets: {
+            schemaDoc: 'ID list (for PUSH/EMAIL/SMS) / email list (for Email) / phone list (for SMS), can be mixed',
```
Comment (AntonAL) [inline apps/condo/domains/notification/schema/MessageBatch.js:73]:
```suggestion
            schemaDoc: 'List of ids for "push", "email", "sms" message types; list of emails for "email" message types; list of phone numbers for "sms" message types. Can be mixed. For each entry an appropriate message type will be used.',
```

File: apps/condo/domains/notification/schema/MessageBatch.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+                    if (!isArray(targets) || isEmpty(targets)) addFieldValidationError(`${JSON_EXPECT_ARRAY_ERROR}${fieldPath}] Expect non-empty JSON Array`)
+
+                    operationForbiddenValidator(args)
+                },
+            },
+        },
+
+        status: {
+            schemaDoc: 'Batch processing status',
```
Comment (AntonAL) [inline apps/condo/domains/notification/schema/MessageBatch.js:89]:
```suggestion
            schemaDoc: 'Status of processing sending messages task',
```

File: apps/condo/domains/notification/schema/MessageBatch.js (hunk 6 above)
Comment (akarjakin) [reply apps/condo/domains/notification/schema/MessageBatch.js:32]: Actually it does. This is batch, it is handled by async task.

File: apps/condo/domains/common/utils/sleep.js (hunk 5 above)
Comment (akarjakin) [reply apps/condo/domains/common/utils/sleep.js:6]: That should be enough I think. Every time it fails on 2999 for 3000 and so on. Looks like it laсks a fraction of second.

**Comment by akarjakin** [comment]
> Is it really convenient to enter data by hands from console when using `send-custom-message-batch.js` script ?
> 
> Maybe to load from YAML file of strict format.
> 
> For example:
> 
> ```
> message:
>   title: Hello
>   body: |
>   	This is a custom message
>   	With multiple lines
>   deepLink: https://example.com/#hello
>   phones: |
>   	+79991111111
>   	+79992222222
>   	+79993333333
>   emails: |
>     anton@example.com
>     andrew@example.com
>     pavel@example.com
> ```
> 
> This way it's more easier to work, because this file can be prepared by other person, you can store it somewhere, send to somebody etc.

There is a problem though - how will you transfer that YAML file to production pod? :)

**Comment by akarjakin** [comment]
> A function `connectKeystone` is duplicated in 4 places. Maybe it's time to extract and reuse it?

I'm going to fix some scripts very soon, I'd prefer to refactor that there.

File: apps/condo/domains/notification/schema/MessageBatch.js (hunk 6 above)
Comment (AntonAL) [reply apps/condo/domains/notification/schema/MessageBatch.js:32]: MessageBatch is a noun, it's primary side-effect is an action — verb.
There is no verb in name.

**Review by AntonAL — APPROVED** [review]

File: apps/condo/domains/notification/tasks/index.js [hunk 12]
```diff
@@ -1,5 +1,5 @@
-const { sendRemoteClientsUpgradeAppNotifications } = require('./sendRemoteClientsUpgradeAppNotifications')
+const { deliverMessage } = require('./deliverMessage')
 
 module.exports = {
-    sendRemoteClientsUpgradeAppNotifications,
```
[the remark this answers is not in the document]
Comment (akarjakin) [reply apps/condo/domains/notification/tasks/index.js:4]: no need

**Pull request merged**